### PR TITLE
Start of python 3 support.

### PR DESCRIPTION
**Pull request by stuaxo**
Installs OK on python 3, need to port the tests.

### CODE REVIEW

File: lib/cmdln.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 # License: MIT (see LICENSE.txt for license details)
 # Author:  Trent Mick
 
+from __future__ import print_function
```
Comment (trentm) [inline lib/cmdln.py]: can we move that to after the module docstring? else `__doc__` doesn't work.

File: lib/cmdln.py [hunk 2]
```diff
@@ -237,13 +240,13 @@ class MyCmd(cmdln.Cmdln):
         if self.optparser: # i.e. optparser=None means don't process for opts
             try:
                 self.options, args = self.optparser.parse_args(argv[1:])
-            except CmdlnUserError, ex:
+            except CmdlnUserError as ex:
```
Comment (trentm) [inline lib/cmdln.py]: Hrm, this breaks python 2.4 and 2.5 compat. I'm not sure how important that is. Definitely means a major rev bump.

**Comment by stuaxo** [comment]
Ah, probably worth a major version bump.  

I moved the from __future__ import to later (I doubt earlier than 2.7 has it?).

There are a lot of things that stop the tests passing... I can fix the obvious bits with exceptions, there are then a whole bunch of things where it's confused between unicode and bytes - stderr and stdout return bytes by default, so you have to encode to unicode IRCC..    as I've don't know this code at all, not sure what the correct fix is.

Definitely want to use this in my py3 project since it's one of the only sane looking commandline handlers :)

**Comment by trentm** [comment]
Otherwise looks like a fine start. Ugh I haven't released cmdln to pypi in quite a while, and the tests are failing. This might need some love. Would be nice to get 1.3.0 publisyhed before bumping to 2.0.0 for this change.

**Comment by trentm** [comment]
Work now. Hopefully I can take a look tonight at the other pieces.

**Comment by stuaxo** [comment]
Cool :)  best thing is just to make a seperate python3 virtualenv then you can test in both.

**Comment by stuaxo** [comment]
In cmdline, round line 649:

```python
        cmdnames = {} # use a dict to strip duplicates
        for attr in self.get_names():
            if attr.startswith("do_"):
                cmdnames[attr[3:]] = True
        cmdnames = cmdnames.keys()
        cmdnames.sort()
        linedata = []
        for cmdname in cmdnames:
            if aliases.get(cmdname):
```

This gets
```AttributeError: 'dict_keys' object has no attribute 'sort'```

It probably makes sense to use a set here, if bumping to 2.7 then something like this should work:

```python
        cmdnames = set()
        for cmdname in [ attr[3:] for attr in self.get_names() if attr.startswith("do_")]:
            cmdnames.add(cmdname)
        linedata = []
        for cmdname in sorted(cmdnames):
            if aliases.get(cmdname):
```

**Comment by stuaxo** [comment]
Next things I came up with:

In ```cmdln.py```  Search and replace 

```.im_func``` with ```.__func__``` (find from py2.6 - 3.0)
Also 
```im_func``` with ```__code__``` though I'm not sure how old ```__code__``` goes...  but doing those it seems like the project I'm writing might even work..

**Comment by trentm** [comment]
I've brought the repo to the present (cleaned out some cruft, issues, pulls, done a 1.3.x release to pypi). The master branch is now for python 3 support, so we can iterate on that. I'll try to help you but I don't have much bandwidth at all :) and I don't currently using cmdln a whole lot with python 3 yet.

**Comment by trentm** [comment]
Merged in commit 1740f659f13803363283c107da9f1fbbd4c3db7f.

Shall we move to new PRs for further work?

`make test` will run the test suite now using whatever `python` is on your PATH.  Use `make test34` to run with python3.4. `make testall` to test with 2.6, 2.7, 3.3 and 3.4 if you have those pythons installed.  Or to test with a particular python: `make test PYTHON=/path/to/a/python`

I haven't looked in depth at the other comments above yet. I'd prefer to maintain 2.6 compat if possible.
Python 2.6 has the `print_function`: https://docs.python.org/2/library/__future__.html

**Comment by trentm** [comment]
> In cmdline, round line 649:

I took your fix for this (worked in python 2.6 too): commit ba5bb09

**Comment by trentm** [comment]
> __func__ ... __code__

Thanks!  Took those in commit 34aab7a.

**Comment by trentm** [comment]
Next problem is this:

```
$ python3.4 docs/conan.py crush -h
Traceback (most recent call last):
  File "docs/conan.py", line 52, in <module>
    sys.exit( conan.main() )
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 264, in main
    return self.cmd(args)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 287, in cmd
    retval = self.onecmd(argv)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 425, in onecmd
    return self._dispatch_cmd(handler, argv)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 1119, in _dispatch_cmd
    opts, args = optparser.parse_args(argv[1:])
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/optparse.py", line 1386, in parse_args
    stop = self._process_args(largs, rargs, values)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/optparse.py", line 1430, in _process_args
    self._process_short_opts(rargs, values)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/optparse.py", line 1535, in _process_short_opts
    option.process(opt, value, values, self)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/optparse.py", line 784, in process
    self.action, self.dest, opt, value, values, parser)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/optparse.py", line 806, in take_action
    parser.print_help()
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 1016, in print_help
    self.cmdln.onecmd(["help", self.subcmd])
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 425, in onecmd
    return self._dispatch_cmd(handler, argv)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 1110, in _dispatch_cmd
    return handler(argv)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 532, in do_help
    doc = self._help_preprocess(doc, cmdname)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 613, in _help_preprocess
    help = preprocessor(help, cmdname)
  File "/Users/trentm/tm/cmdln/lib/cmdln.py", line 769, in _help_preprocess_cmd_usage
    if func.func_defaults:
AttributeError: 'function' object has no attribute 'func_defaults'
```

**Comment by trentm** [comment]
Fixed that in commit c67ac34

**Comment by trentm** [comment]
Now the test suite with python 3.x hangs on the process handling from "test/process.py".

```
$ make test33
...
^C (to cancel)
  File "/Users/trentm/tm/cmdln/test/test_cmdln.py", line 285, in _testOneCmdln
    retval = p.wait()
  File "/Users/trentm/tm/cmdln/test/process.py", line 1400, in wait
    pid, sts = os.waitpid(self._pid, timeout)
KeyboardInterrupt
make[1]: *** [test] Error 1
make: *** [test33] Interrupt: 2
```

This might be gross. process.py is a process handling lib I wrote back in ancient days before python had the subprocess module. For a while it handled somethings subprocess.py didn't. Unlikely that's still true or that it'll matter here. We might want to change to subprocess.py.  Note that the cmdln test suite is doing crazy exec'ing of `tclsh` (!) to do expect (http://www.tcl.tk/man/expect5.31/expect.1.html) -based testing.  Not sure if that'll get in the way.

I'm done banging on this for tonight.

**Comment by stuaxo** [comment]
It might be worth switching to pexpect for the 'expect' bits, it looks like version 4 should actually support windows too.

https://github.com/pexpect/pexpect/issues/17

**Comment by trentm** [comment]
Got it working... the problem was process.py, not expect usage. Works fine with subprocess. After a few more changes the test suite is fully passing in python 3.3 and 3.4!  I've pushed 2.0 to pypi. Let me know if you hit python 3 issues.
Thanks,
Trent

**Comment by stuaxo** [comment]
Great work, have upgraded and at first glance seems to work, will certainly let you know.

**Pull request closed without merging**